src/ra_server.erl

```diff
@@ -2818,6 +2889,99 @@ meta_name(#cfg{system_config = #{names := #{log_meta := Name}}}) ->
     Name;
 meta_name(#{names := #{log_meta := Name}}) ->
     Name.
+
+already_member(State) ->
+    % already a member do nothing
+    % TODO: reply? If we don't reply the caller may block until timeout
+    {not_appended, already_member, State}.
+
+%%% ====================
+%%% Voter status helpers
+%%% ====================
+
+-spec ensure_promotion_target(ra_voter_status(), ra_server_state()) ->
+    {ok, ra_voter_status()} | {error, term()}.
+ensure_promotion_target({voter, Reason}, _) ->
+    {ok, {voter, Reason}};
+ensure_promotion_target({nonvoter, #{target := _, nvid := _} = Reason}, _) ->
+    {ok, {nonvoter, Reason}};
+ensure_promotion_target({nonvoter, #{nvid := _} = Reason}, #{commit_index := CI}) ->
+    {ok, {nonvoter, Reason#{target => CI}}};
+ensure_promotion_target(_, _) ->
+    {error, missing_nvid}.
+
+-spec init_voter_status(ra_server_config() | ra_new_server()) -> ra_voter_status().
+init_voter_status(#{init_non_voter := NVId}) ->
+    {nonvoter, #{nvid => NVId}};
+init_voter_status(_) ->
+    {voter, #{}}.
+
+-spec voter_status(ra_server_state()) -> ra_voter_status().
+voter_status(#{cluster := Cluster} = State) ->
+    case maps:get(voter_status, State, undefined) of
+        undefined ->
+            voter_status(id(State), Cluster);
+        Voter ->
+            Voter
+    end.
+
+-spec voter_status(ra_server_id(), ra_cluster()) -> ra_voter_status().
+voter_status(PeerId, Cluster) ->
+    case maps:get(PeerId, Cluster, undefined) of
+        undefined ->
+            {undefined, #{}};
+        Peer ->
+            maps:get(voter_status, Peer, {voter, #{}})
+    end.
+
+-spec maybe_promote_self(ra_cluster(), ra_server_state()) -> ra_voter_status().
+maybe_promote_self(NewCluster, State) ->
+    %% A bit complicated procedure since nvid is not guaranteed.
+    {_, NReason} = New = voter_status(id(State), NewCluster),
+    {_, CReason} = Current = voter_status(State),
+    Self = maps:get(nvid, CReason, undefined),
+    case maps:get(nvid, NReason, undefined) of
+        Self -> New;
+        _ -> Current
+    end.
+
+-spec maybe_promote_peer(ra_server_id(), ra_server_state(), effects()) -> effects().
+maybe_promote_peer(PeerID, #{cluster := Cluster}, Effects) ->
+    % Unknown peer handled in the caller.
+    #{PeerID := #{match_index := MI,
+                  voter_status := OldStatus}} = Cluster,
+    case update_voter_status(OldStatus, MI) of
```

Comment:
I think it may be easier to read if the logic form this single use function was inlined here as a pattern match